.github/workflows/lint-and-test.yml

```diff
@@ -12,36 +12,98 @@ name: Lint and run unit tests
 on: [pull_request]
 
 jobs:
+  setup:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Get list of packages
+        id: set-matrix
+        run: |
+          PACKAGES=$(ls -d packages/* | jq -R -s -c 'split("\n")[:-1]')
+          echo "Found packages: $PACKAGES"
+          echo "::set-output name=matrix::${PACKAGES}"
+
   test:
+    needs: setup
     runs-on: ubuntu-20.04
-
     strategy:
+      fail-fast: false
       matrix:
-        node-version: [16.20.0]
+        package: ${{fromJson(needs.setup.outputs.matrix)}}
 
     steps:
       - name: Checking out git repo
         uses: actions/checkout@v2
 
+      - name: Check package.json and scripts
+        id: check-scripts
+        run: |
+          if [ ! -f "${{ matrix.package }}/package.json" ]; then
+            echo "No package.json found for ${{ matrix.package }}. Stopping pipeline."
+            echo "::set-output name=skip::true"
+          else
+            echo "::set-output name=skip::false"
+
+            if ! grep -q "\"test\":" "${{ matrix.package }}/package.json"; then
+              echo "Test not found in ${{ matrix.package }}"
+              echo "::set-output name=skip-test::true"
+            else
+              echo "::set-output name=skip-test::false"
+            fi
+
+            if ! grep -q "\"lint\":" "${{ matrix.package }}/package.json"; then
+              echo "Lint scripts not found in ${{ matrix.package }}. Stopping pipeline."
+              echo "::set-output name=skip-lint::true"
+            else
+              echo "::set-output name=skip-lint::false"
+            fi
+          fi
+
       - name: Use Node.js ${{ matrix.node-version }}
+        if: steps.check-scripts.outputs.skip != 'true'
         uses: actions/setup-node@v2
         with:
           node-version: ${{ matrix.node-version }}
           cache: 'npm'
 
+      - name: Remove other package directories
```

Comment:
I added this so that each runner could install only the dependencies the specific package needs. Makes the install step a lot faster.